treeherder/perf/migrations/0010_fix_signature_uniqueness.py

```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.13 on 2018-09-04 09:38
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('model', '0008_remove_failure_match'),
+        ('perf', '0009_non_nullable_issue_tracker'),
+    ]
+
+    operations = [
+        migrations.AlterUniqueTogether(
+            name='performancesignature',
```

Comment:
How big is this table? (Time taken to run the migration)

Could we try the migration on prototype? Also worth noting that there's presumably a chance the constraint could fail when it runs on production, due to it not having been enforced before.

Reply:
> How big is this table? (Time taken to run the migration)
I couldn't say something about the duration of running this. ATM the performance_signature table contains around 350K rows.

> Could we try the migration on prototype?
Yes, we should that.

> Also worth noting that there's presumably a chance the constraint could fail when it runs on production, due to it not having been enforced before.
How easy is it for us to recover from a deployment exception?

Reply:
> How easy is it for us to recover from a migration exception during deployment?

I *think* in this case the new indexes/constraints would just get rolled back, but you'd need to check the MySQL docs or with a DBA to be sure.

Reply:
> ATM the performance_signature table contains around 350K rows.

Ah small enough that time taken won't be an issue (guessing seconds to minutes).

Reply:
I found this [link](https://dev.mysql.com/doc/refman/5.7/en/cannot-roll-back.html) mentioning DDL statements that could not be rolled back. Unfortunately, it doesn't mention about ALTER TABLE ADD UNIQUE CONSTRAINT specifically.

Reply:
Found this in the [MySQL docs](https://dev.mysql.com/doc/refman/5.7/en/innodb-online-ddl-operations.html):

> When you create a UNIQUE or PRIMARY KEY index, MySQL must do some extra work. For UNIQUE indexes, MySQL checks that the table contains no duplicate values for the key.

This sounds like we would only get a console error, but the database shouldn't have problems (schema & data wise).  This person [here](https://dev.mysql.com/doc/refman/8.0/en/alter-table-examples.html#c12143) confirms my assumption, saying that and error message will be prompted if duplicated data already exists.